File: leetcode/p00349/p00349.go
package p00349

// Time Complexity: O(n)
// Space Complexity: O(n)
func intersection(nums1 []int, nums2 []int) []int {
	nums := nums1
	stream := nums2
	if len(nums2) < len(nums1) {
		nums = nums2
		stream = nums1
	}

	table := map[int]bool{}
	for _, n := range nums {
		table[n] = true
	}

	res := []int{}
	for i := 0; i < len(stream); i++ {
		if unvisited, exists := table[stream[i]]; exists && unvisited {
			res = append(res, stream[i])
			table[stream[i]] = false
		}
	}

	return res
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func intersection_sol2(nums1 []int, nums2 []int) []int {
	table1 := map[int]int{}
	table2 := map[int]int{}

	for _, n := range nums1 {
		table1[n]++
	}
	for _, n := range nums2 {
		table2[n]++
	}

	res := []int{}
	for k := range table1 {
		if _, exists := table2[k]; exists {
			res = append(res, k)
		}
	}

	return res
}

File: leetcode/p00349/p00349.py
from typing import List

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        return list(set(nums1) & set(nums2))

if __name__ == '__main__':
    s = Solution()

    assert sorted(s.intersection([1,2,2,1],[2,2])) == [2]
    assert sorted(s.intersection([4,9,5],[9,4,9,8,4])) == [4,9]